corehq/apps/userreports/management/commands/prune_old_datasources.py

```diff
@@ -29,67 +31,80 @@ def add_arguments(self, parser):
-                  AND (
-                    table_name LIKE '{UCR_TABLE_PREFIX}%%'
-                    OR
-                    table_name LIKE '{LEGACY_UCR_TABLE_PREFIX}%%'
-                );
-                """).fetchall()
-                tables_in_db = {r[0] for r in results}
-
-            tables_to_remove_by_engine[engine_id] = tables_in_db - expected_tables
-
-        for engine_id, tablenames in tables_to_remove_by_engine.items():
-            print("\nTables no longer referenced in database: {}:\n".format(engine_id))
-            engine = connection_manager.get_engine(engine_id)
-            if not tablenames:
-                print("\t No tables to prune")
-                continue
-
-            for tablename in tablenames:
-                with engine.begin() as connection:
-                    try:
-                        result = connection.execute(f'SELECT COUNT(*), MAX(inserted_at) FROM "{tablename}"')
-                    except Exception:
-                        print(f"\t{tablename}: no inserted_at column, probably not UCR")
-                    else:
-                        row_count, idle_since = result.fetchone()
-                        if row_count == 0:
-                            print(f"\t{tablename}: {row_count} rows")
-                            if options['drop_empty_tables']:
-                                connection.execute(f'DROP TABLE "{tablename}"')
-                                print(f'\t^-- deleted {tablename}')
-                        else:
-                            print(f"\t{tablename}: {row_count} rows, idle since {idle_since}")
+def prune_tables(tables_to_remove_by_engine, drop_empty_tables, drop_deleted_tables):
+    deleted_domains = Domain.get_deleted_domain_names()
+    for engine_id, tablenames in tables_to_remove_by_engine.items():
+        print("\nTables no longer referenced in database: {}:\n".format(engine_id))
+        engine = connection_manager.get_engine(engine_id)
+        if not tablenames:
+            print("\t No tables to prune")
+            continue
 
+        for tablename in tablenames:
+            should_delete = drop_deleted_tables and get_domain_for_ucr_table_name(tablename) in deleted_domains
 
-
-    def _get_tables_by_engine_id(self, data_sources, engine_id):
-        tables_by_engine_id = defaultdict(set)
-        for data_source in data_sources:
-            if engine_id and data_source.engine_id != engine_id:
-                continue
-
-            table_name = get_table_name(data_source.domain, data_source.table_id)
-            tables_by_engine_id[data_source.engine_id].add(table_name)
-
-        return tables_by_engine_id
+            with engine.begin() as connection:
+                try:
+                    result = connection.execute(f'SELECT COUNT(*), MAX(inserted_at) FROM "{tablename}"')
+                except Exception:
+                    print(f"\t{tablename}: no inserted_at column, probably not UCR")
```

Comment:
I know this isn't your code, but that message seems presumptuous. It might be better if we raised that for the specific db error corresponding to the missing column and had a more generic message for other exceptions.

Reply:
So I did look into this before from the angle of trying to catch a more specific exception, but had trouble when testing in a shell. The error raised is a ProgrammingError, with the underlying exception being `psycopg2.errors.UndefinedColumn`. When trying to catch a `django.db.ProgrammingError` instead of a general `Exception`, the error is not properly caught. Any ideas why that would be the case?

Reply:
Cal flagged the other ProgrammingError from sqlalchemy, and that resolved this issue. See 125fd0d449a47e772958bf20ea3886737e46bdd2